src/Xrefcheck/Verify.hs

```diff
@@ -138,6 +139,35 @@ instance Buildable VerifyError where
         [h] -> ",\n   did you mean " +| h |+ "?\n"
         hs  -> ", did you mean:\n" +| blockListF' "    -" build hs
 
+data NeedsCaching key
+  = NoCaching
+  | CacheUnderKey key
+
+forConcurrentlyCaching
+  :: Ord cacheKey
+  => [a] -> (a -> NeedsCaching cacheKey) -> (a -> IO b) -> IO [b]
+forConcurrentlyCaching = go [] M.empty
+  where
+    go
+      :: Ord cacheKey
+      => [Async b]
+      -> M.Map cacheKey (Async b)
+      -> [a]
+      -> (a -> NeedsCaching cacheKey)
+      -> (a -> IO b)
+      -> IO [b]
+    go acc cached (x : xs) needsCashing action = case needsCashing x of
+      NoCaching -> do
+        withAsync (action x) $ \b ->
+          go (b : acc) cached xs needsCashing action
+      CacheUnderKey cacheKey -> do
+        case M.lookup cacheKey cached of
+          Nothing -> do
+            withAsync (action x) $ \b ->
+              go (b : acc) (M.insert cacheKey b cached) xs needsCashing action
+          Just b -> go (b : acc) cached xs needsCashing action
+    go acc _ [] _ _ = for acc wait <&> reverse
```

Comment:
Slight improvement suggestion: `needsCashing` and `action` args are not modified in the `go` loop, so they can be "hoisted" to the top-level function, instead of being passed around from one iteration to the next.

```suggestion
forConcurrentlyCaching
  :: forall a b cacheKey
   . Ord cacheKey
  => [a] -> (a -> NeedsCaching cacheKey) -> (a -> IO b) -> IO [b]
forConcurrentlyCaching list needsCashing action = go [] M.empty list
  where
    go
      :: Ord cacheKey
      => [Async b]
      -> M.Map cacheKey (Async b)
      -> [a]
      -> IO [b]
    go acc cached (x : xs) = case needsCashing x of
      NoCaching -> do
        withAsync (action x) $ \b ->
          go (b : acc) cached xs
      CacheUnderKey cacheKey -> do
        case M.lookup cacheKey cached of
          Nothing -> do
            withAsync (action x) $ \b ->
              go (b : acc) (M.insert cacheKey b cached) xs
          Just b -> go (b : acc) cached xs
    go acc _ [] = for acc wait <&> reverse
```

Reply:
Will do.
I've also been thinking about the fact that `NeedsCaching` is equivalent to `Maybe` and perhaps, even though it is descriptive, is redundant and could be replaced with `Maybe`. @Martoon-00, what do you think?

Reply:
> even though it is descriptive

Personally, I think this is a good enough reason to keep it. It improves readability for maintainers in the future.